ff_dir.c

```diff
@@ -2941,160 +2949,165 @@ FF_Error_t FF_CreateDirent( FF_IOManager_t * pxIOManager,
+
                 {
-                    /* Hash the directory. */
-                    FF_HashDir( pxIOManager, ulDirCluster );
+                    FF_Error_t xTempError;
+
+                    xTempError = FF_CleanupEntryFetch( pxIOManager, &xFetchContext );
+
+                    if( FF_isERR( xReturn ) == pdFALSE )
+                    {
+                        xReturn = xTempError;
+                    }
                 }
 
-                memcpy( pcShortName, pucEntryBuffer, 11 );
-                FF_ProcessShortName( pcShortName ); /* Format the shortname to 8.3. */
-                #if ( ffconfigHASH_FUNCTION == CRC16 )
+                if( FF_isERR( xReturn ) )
                 {
-                    FF_AddDirentHash( pxIOManager, ulDirCluster, ( uint32_t ) FF_GetCRC16( ( uint8_t * ) pcShortName, ( uint32_t ) strlen( pcShortName ) ) );
+                    break;
                 }
-                #elif ( ffconfigHASH_FUNCTION == CRC8 )
+
+                #if ( ffconfigHASH_CACHE != 0 )
                 {
-                    FF_AddDirentHash( pxIOManager, ulDirCluster, ( uint32_t ) FF_GetCRC8( ( uint8_t * ) pcShortName, strlen( pcShortName ) ) );
+                    if( FF_DirHashed( pxIOManager, ulDirCluster ) == pdFALSE )
+                    {
+                        /* Hash the directory. */
+                        FF_HashDir( pxIOManager, ulDirCluster );
+                    }
+
+                    memcpy( pcShortName, pucEntryBuffer, 11 );
+                    FF_ProcessShortName( pcShortName ); /* Format the shortname to 8.3. */
+                    #if ( ffconfigHASH_FUNCTION == CRC16 )
+                    {
+                        FF_AddDirentHash( pxIOManager, ulDirCluster, ( uint32_t ) FF_GetCRC16( ( uint8_t * ) pcShortName, ( uint32_t ) strlen( pcShortName ) ) );
+                    }
+                    #elif ( ffconfigHASH_FUNCTION == CRC8 )
+                    {
+                        FF_AddDirentHash( pxIOManager, ulDirCluster, ( uint32_t ) FF_GetCRC8( ( uint8_t * ) pcShortName, strlen( pcShortName ) ) );
+                    }
+                    #endif /* ffconfigHASH_FUNCTION */
                 }
-                #endif /* ffconfigHASH_FUNCTION */
+                #endif /* ffconfigHASH_CACHE*/
             }
-            #endif /* ffconfigHASH_CACHE*/
         }
-    }
-    while( pdFALSE );
+        while( pdFALSE );
 
-    FF_UnlockDirectory( pxIOManager );
+        FF_UnlockDirectory( pxIOManager );
 
-    if( FF_isERR( xReturn ) == pdFALSE )
+        if( FF_isERR( xReturn ) == pdFALSE )
+        {
+            if( pxDirEntry != NULL )
+            {
+                pxDirEntry->usCurrentItem = ( uint16_t ) ( lFreeEntry + xLFNCount );
+            }
+        }
+    } /* if( FF_MakeNameCompliant( pxDirEntry->pcFileName ) ) */
+    else
     {
-        if( pxDirEntry != NULL )
+        if( pxDirEntry->ucAttrib & FF_FAT_ATTR_DIR )
```

Comment:
Sorry one more remark related to MISRA:

~~~diff
-    if( pxDirEntry->ucAttrib & FF_FAT_ATTR_DIR )
+     if( ( pxDirEntry->ucAttrib & FF_FAT_ATTR_DIR ) != 0U )
~~~
This test (`!= 0U`) will make it a real boolean value.

Reply:
Okay